data/glsk/glsk-document-cse/src/main/java/com/farao_community/farao/data/glsk/cse/CseGlskShiftKey.java

```diff
@@ -0,0 +1,102 @@
+        initCommonMemberVariables(glskBlockElement, businessType, pointInterval, subjectDomainmRID);
+
+        if ("ManualGSKBlock".equals(glskBlockElement.getTagName())) {
+            this.businessType = "B43";
+            NodeList nodesList = glskBlockElement.getElementsByTagName("Node");
+            double currentFactorsSum = 0;
+            for (int i = 0; i < nodesList.getLength(); i++) {
+                Element nodeElement = (Element) nodesList.item(i);
+                CseGlskRegisteredResource cseRegisteredResource = new CseGlskRegisteredResource(nodeElement);
+                registeredResourceArrayList.add(cseRegisteredResource);
+                Optional<Double> initialFactor = cseRegisteredResource.getInitialFactor();
+                if (initialFactor.isPresent()) {
+                    currentFactorsSum += initialFactor.get();
+                }
+            }
+
+            if (currentFactorsSum == 0) {
+                throw new GlskException("Factors sum should not be 0");
+            }
+
+            for (AbstractGlskRegisteredResource registeredResource : registeredResourceArrayList) {
+                CseGlskRegisteredResource cseRegisteredResource = (CseGlskRegisteredResource) registeredResource;
+                Optional<Double> intialFactor = cseRegisteredResource.getInitialFactor();
+                if (intialFactor.isPresent()) {
+                    cseRegisteredResource.setParticipationFactor(intialFactor.get() / currentFactorsSum);
+                }
+            }
+        } else if ("PropGSKBlock".equals(glskBlockElement.getTagName())) {
+            importImplicitProportionalBlock(glskBlockElement, "B42");
+        } else if ("PropLSKBlock".equals(glskBlockElement.getTagName())) {
+            this.psrType = "A05"; // Enforce psrType that does not respect "official" format specification
+            importImplicitProportionalBlock(glskBlockElement, "B42");
+        } else if ("ReserveGSKBlock".equals(glskBlockElement.getTagName())) {
+            importImplicitProportionalBlock(glskBlockElement, "B44");
+        } else {
+            throw new GlskException("Unknown UCTE Block type");
+        }
+    }
+
+    public CseGlskShiftKey(Element glskBlockElement, String businessType, Interval pointInterval, String subjectDomainmRID, int position) {
+        initCommonMemberVariables(glskBlockElement, businessType, pointInterval, subjectDomainmRID);
+        this.meritOrderPosition = position;
+
+        if ("MeritOrderGSKBlock".equals(glskBlockElement.getTagName())) {
+            this.businessType = "B45";
+            Element nodeElement = (Element) glskBlockElement.getElementsByTagName("Node").item(position);
+            CseGlskRegisteredResource cseRegisteredResource = new CseGlskRegisteredResource(nodeElement);
+            registeredResourceArrayList.add(cseRegisteredResource);
+        } else {
+            throw new GlskException("Unknown UCTE Block type");
+        }
+    }
```

Comment:
I feel like this could be simplified. In the end you always add a registered resource with a different business type, i don't feel like you need two different constructors and then two different methods in the caller class. I get that there is a loop for merit order and not for the others, but it feels a bit tricky to understand here for something that is not really complicated.

Reply:
I think it is better to do it that way compared to provide a fake position for other type of GLSK that won't be used.

Reply:
And what if you add the case MeritOrderGskBlock in the first constructor and you do the loop inside this case as it is done for ManualGSKBlock ?

Reply:
The situation is a bit different, cause there are multiple shift keys created, and the loop index has an impact on the created glsk shift key, but is not explicited in the document (the merit order position) contrary to CIM for example. What you propose does not work, from my point of view.

Reply:
Ok I get it, I didn't notice the difference of the loop between manual gsk and merit order glsk